src/cli_utils.cc

```diff
@@ -78,69 +78,143 @@ Status LoadYAMLFile(const string& filename, YAML::Node &node) {
+        contigs.push_back(make_pair(name, size));
+    }
 
-    yaml << YAML::BeginMap;
+    return Status::OK();
+}
+
+// We write the stream as follows:
+//
+// ---
+// contigs
+// ---
+// allele 1
+// ---
+// allele 2
+// ---
+// etc.
+// allele N
+// ...
+//
+// Each element is transformed to YAML, and then written to the output stream.
+// This generates the document in pieces, while keeping it valid YAML.
+Status yaml_stream_of_discovered_alleles(const std::vector<std::pair<std::string,size_t> > &contigs,
+                                         const discovered_alleles &dsals,
+                                         std::ostream &os) {
+    Status s;
 
-    // write contigs
-    yaml << YAML::Key << "contigs";
-    yaml << YAML::Value;
+    // write the contigs
     {
-        yaml << YAML::BeginSeq;
-        for (const std::pair<string,size_t> &pr : contigs) {
-            yaml << YAML::BeginMap;
-            yaml << YAML::Key << "name";
-            yaml << YAML::Value << pr.first;
-            yaml << YAML::Key << "size";
-            yaml << YAML::Value << pr.second;
-            yaml << YAML::EndMap;
-        }
-        yaml << YAML::EndSeq;
+        os << "---" << endl;
+        YAML::Emitter yaml;
+        S(yaml_of_contigs(contigs, yaml));
+        os << yaml.c_str() << endl;
     }
 
     // Write alleles
-    yaml << YAML::Key << "alleles";
-    yaml << YAML::Value;
-    {
-        // Note: we might write out empty sites too
-        s = yaml_of_discovered_alleles(dsals, contigs, yaml);
-        if (s.bad()) return s;
+    for (auto &pr : dsals) {
+        os << "---" << endl;
+        YAML::Emitter yaml;
+        S(yaml_of_one_discovered_allele(pr.first, pr.second, contigs, yaml));
+        os << yaml.c_str() << endl;
     }
 
-    yaml << YAML::EndMap;
+    // special notation for end-of-file
+    os << "..." << endl;
 
     return Status::OK();
 }
 
-Status contigs_alleles_of_yaml(const YAML::Node& yaml,
-                               std::vector<std::pair<std::string,size_t> > &contigs,
-                               discovered_alleles &dsals) {
-    Status s;
-    if (!yaml.IsMap()) {
-        return Status::Invalid("not a map at top level");
+
+// In a YAML document built out of top-level sub-documents, get the next document.
+//
+// Notes:
+// -  Catch any exceptions, and return an empty node.
+static YAML::Node yaml_get_next_document(std::istream &is, bool skip_first_line = false) {
+    try {
+        char buf[200];
+        stringstream ss;
+        if (skip_first_line) {
+            // the document starts with a "---", skip it
+            is.getline(buf, 200);
+        }
+
+        while (is.good() && !is.eof()) {
+            is.getline(buf, 200);
+            size_t buf_len = strlen(buf);
+
+            if (buf_len == 3) {
+                // check if we reached the end of this top level document
+                string marker(buf);
+                if (marker == "---" || marker == "...")
```

Comment:
It would be great to error out in some way if we hit EOF without seeing the "..."